### PR TITLE
CTB-881 close bucket writers that fail to flush

### PR DESCRIPTION
**Pull request by tempredirect**
It appears to be possible that a bucket writer can get it's self into a
state where it accepts appends but fails the flush. It's difficult to
pin down exactly the circumstances of what causes this, but it
seems reasonable to perform a close on the bucket writer on failure
in a similar way to the append method does.

Note the bucket writer is unable to implement this directly as
it is close calls flush so it'd end in a loop.

### CODE REVIEW

**Comment by navidn78** [comment]
+1

File: flume-ng-sinks/flume-hdfs-sink/src/main/java/org/apache/flume/sink/hdfs/HDFSEventSink.java [hunk 1]
```diff
@@ -475,6 +472,28 @@ public void run(String bucketPath) {
     }
   }
 
+  private void flushWriters(List<BucketWriter> writers) throws InterruptedException, IOException {
+    // flush all pending buckets
+    // close any that error and rethrow the first exception
```
Comment (jakecollins) [inline flume-ng-sinks/flume-hdfs-sink/src/main/java/org/apache/flume/sink/hdfs/HDFSEventSink.java]: seems to rethrow the last exception not the first?

File: flume-ng-sinks/flume-hdfs-sink/src/test/java/org/apache/flume/sink/hdfs/HDFSTestSeqWriter.java [hunk 2]
```diff
@@ -52,9 +54,12 @@ public void append(Event e) throws IOException {
       throw new IOException("Injected fault");
     } else if (e.getHeaders().containsKey("fault-until-reopen")) {
       // opening first time.
-      if(openCount == 1) {
```
Comment (jakecollins) [inline flume-ng-sinks/flume-hdfs-sink/src/test/java/org/apache/flume/sink/hdfs/HDFSTestSeqWriter.java]: What's going on here? Perhaps add comments to explain?

Comment (tempredirect) [inline flume-ng-sinks/flume-hdfs-sink/src/test/java/org/apache/flume/sink/hdfs/HDFSTestSeqWriter.java]: I didn't write that

File: flume-ng-sinks/flume-hdfs-sink/src/main/java/org/apache/flume/sink/hdfs/HDFSEventSink.java (hunk 1 above)
Comment (tempredirect) [inline flume-ng-sinks/flume-hdfs-sink/src/main/java/org/apache/flume/sink/hdfs/HDFSEventSink.java]: see latest

**Pull request merged**